libraries/botbuilder-dialogs-adaptive-teams/src/actions/baseTeamsCacheInfoResponseDialog.ts

```diff
@@ -0,0 +1,125 @@
+     */
+    public disabled: BoolExpression;
+
+    /**
+     * Gets or sets config CacheType.
+     *
+     * @example
+     * "cache" or "no_cache".
+     */
+    public cacheType: StringExpression;
+
+    /**
+     * Gets or sets cache duration in seconds for which the cached object should remain in the cache.
+     */
+    public cacheDuration: IntExpression;
+
+    public getConverter(
+        property: keyof BaseTeamsCacheInfoResponseDialogConfiguration | string
+    ): Converter | ConverterFactory {
+        switch (property) {
+            case 'disabled':
+                return new BoolExpressionConverter();
+            case 'cacheType':
+                return new StringExpressionConverter();
+            case 'cacheDuration':
+                return new IntExpressionConverter();
+            default:
+                return super.getConverter(property);
+        }
+    }
+
+    protected static createInvokeResponseActivity(
+        body: MessagingExtensionResponse,
+        statusCode: number = StatusCodes.OK
+    ): Activity {
+        return <Activity>{
+            value: {
+                status: statusCode,
+                body: body,
+            },
+            type: ActivityTypes.InvokeResponse,
+        };
+    }
+
+    protected createMessagingExtensionInvokeResponseActivity(
+        dc: DialogContext,
+        result: MessagingExtensionResult
+    ): Activity {
+        switch (dc.context.activity.name) {
+            case 'composeExtension/queryLink':
+            case 'composeExtension/query':
+            case 'composeExtension/selectItem':
+            case 'composeExtension/querySettingsUrl':
+                return BaseTeamsCacheInfoResponseDialog.createInvokeResponseActivity(<MessagingExtensionResponse>{
+                    composeExtension: result,
+                    cacheInfo: this.getCacheInfo(dc),
+                });
+            case 'composeExtension/submitAction':
+            case 'composeExtension/fetchTask':
+                return BaseTeamsCacheInfoResponseDialog.createInvokeResponseActivity(<MessagingExtensionActionResponse>{
+                    composeExtension: result,
+                    cacheInfo: this.getCacheInfo(dc),
+                });
+            default:
+                throw new Error(`GetMessagingExtensionResponse Invalid activity.name: ${dc.context.activity.name}`);
+        }
+    }
+
+    protected getCacheInfo(dc: DialogContext): CacheInfo {
+        if (this.cacheType != null && this.cacheDuration != null) {
+            const cacheType = this.cacheType?.getValue(dc.state);
+            let cacheDuration = this.cacheDuration?.getValue(dc.state);
+
+            if (cacheDuration > 0 && cacheType != null) {
+                // Valid ranges for cacheDuration are 60 < > 2592000
+                cacheDuration = Math.min(Math.max(60, cacheDuration), 2592000);
+                return <CacheInfo>{
```

Comment:
You should rarely need to do this kind of casting. The compiler is smart enough to know that if a function returns type T that any time you type `return foo`, `foo` ought to be of type `T`. When you write `<T>{ ... }`, you perform typecasting, which disables compiler checks. This is almost always a code smell.